### PR TITLE
Updates to the release strategy

### PR DESCRIPTION
**Pull request by mattcaswell**
Remove out of date sections, and add some words about stability.

In WIP until discussion has taken place.

### CODE REVIEW

File: policies/releasestrat.html
Comment (t-j-h) [inline policies/releasestrat.html]: We need to define what we mean by "public interface" here clearly as there have been rather different views on that.
The two views I have seen regularly expressed:
1) in the header files that are installed
2) in the pod documentation files that are installed

Clearly 2) is a much smaller list than 1).

Comment (mattcaswell) [reply policies/releasestrat.html]: In my mind it is clearly (1)

Comment (tmshort) [reply policies/releasestrat.html]: It's actually a problem that (2) is smaller than (1). Lots of things still need documentation.

Comment (richsalz) [reply policies/releasestrat.html]: FWIW, I remember trying to remove something because it wasn't documented and Tim was the leading proponent of "it's in the header file" but that most of the team agreed and I was definitely in the rough. 
```
; ./util/find-doc-nits -u
# Duplicate i2d_PublicKey in doc/man3/d2i_X509.pod and doc/man3/d2i_PrivateKey.pod
# Found 4577 in util/libcrypto.num
# Found 1607 missing from util/libcrypto.num

# Found 499 in util/libssl.num
# Found 57 missing from util/libssl.num

# Checking macros (approximate)
# Found 229 macros missing (not all should be documented)
; 
```

Comment (richsalz) [reply policies/releasestrat.html]: Try the -d flag also.

Comment (t-j-h) [reply policies/releasestrat.html]: So working those rules taking the function DES_cbc_encrypt as an example - it is in the header file des.h and it is not currently marked deprecated. Do we consider the EVP-level API a replacement?
And in order to remove the API we have to mark it as deprecated first in a release and then wait at least 5 more years before it is removed which can only happen in a API-breaking release itself?

Comment (mattcaswell) [reply policies/releasestrat.html]: Yes. Although we could (potentially) move it to a "libcrypto-legacy" library. The structure could be made opaque (according to the rules stated for making things opaque). And the internal implementation could be changed to call EVP.

File: policies/releasestrat.html
Comment (tmshort) [inline policies/releasestrat.html]: So, for an API to be *modified*, it would have to be documented as deprecated? Not sure that makes sense. The most recent example of a possible API change is to `SSL_set_options()`, where it currently takes and returns a `long`. Changing that to a `uint64_t` allows for a consistent number (64) of options, where as right now, it depends on the compiler defintion of `long`. Obviously, changing the type breaks the ABI, but there's a subtle difference between changing an input parameter to 64bits, and the return value to 64bits. The first could be considered API (source) compatible, since the compiler will automatically extend the input to match function signature. But that doesn't happen for the return value.

Comment (mattcaswell) [reply policies/releasestrat.html]: That's a good point. I was thinking of modifications more along the lines of adding or removing parameters. I'm not sure about changing types.

If you changed SSL_set_options() to take `uint64_t` (which is probably API compatible) - wouldn't that necessarily imply also changing SSL_get_options() to return a `uint64_t` which would be API breaking.
I'd be interested in @vdukhovni's thoughts.

Comment (mattcaswell) [reply policies/releasestrat.html]: I suppose there is also the example of "const"ifying parameters - which we've generally allowed.

Comment (vdukhovni) [reply policies/releasestrat.html]: The discussion here is about the API, not ABI, and the example of widening an integer parameter does not introduce an API break. Existing calls that pass a long continue to work when the target type changes to uint64_t, through the magic of C prototypes and integral promotion rules. Similarly, constification of non-pointer input parameters is harmless.

The goal here is to avoid abrupt API transitions where we have one API in OpenSSL version X and a different incompatible API in version X+1, with no common way to perform some functions across the two versions. The notion of X and X+1 is specified as a time range, since our downstream consumers don't necessarily adopt each and every version, but rather do batch upgrades at major plaform releases, ...  We can debate whether 5 years is "just right", or some other number is better, but the idea should be to provide plenty of opportunity for users doing upgrades to have a real migration path where their software compiles against both the old and the new, and then get to gradually fix the deprecation warnings before the next major upgrade.  That way an OS distribution can rebuild all its packages, and work on modernization of the API without having upgrading OpenSSL blocked on porting all the applications.

File: policies/releasestrat.html
Comment (vdukhovni) [inline policies/releasestrat.html]: I am bit concerned about the "(security fix only mode)" potentially becoming a blocker for upgrades, if some major OS distribution bundles a non-LTS release that goes "(security fix only)" and now needs to support applications running against that, and the release that wants to add the opaque accessors.  We may face this problem with OpenSSL 1.1.0 for example. So I think that adding conditionally enabled macros to header files to support newly required accessors should sometimes also apply to releases that otherwise only get security fixes.

This is in part why my original suggestion explicitly mentioned conditional processing of the new macros based on a compiler flag that indicates that the code is being forward ported to a later API.  An implementation detail, but one that allows us to more leeway to make changes otherwise proscribed for stability reasons.

File: policies/releasestrat.html
Comment (mlindner) [inline policies/releasestrat.html]: Just an openssl user here but renaming functions for naming consistency is something that has been long overdue in OpenSSL and I'm glad it is occurring. We shouldn't stop that from continuing.

**Comment by mlindner** [comment]
The company I work for uses OpenSSL heavily and we're still on 1.0.2 and looking forward to moving to 1.1.1. It's going to force a lot of code rewriting but this will be cleaner code than we had anyway and cause less bugs.

I think the thing to be looking at when making breaking changes is making sure the OpenSSL team answers the question "Are all possible ways this function could be used have a drop in replacement alternative method of performing them? Ideally with little rewriting of code." If that is handled properly then there shouldn't be as much an issue with making breaking changes. Also any breaking changes that are caused should ideally cause compile time errors/warnings rather than hidden subtle breakages. Renaming functions/creating new ones is better than modifying the functionality of existing ones. In fact I'd prefer a strict "never modify functionality of documented LTS release functions, only remove/deprecate them".

File: policies/releasestrat.html
Comment (vdukhovni) [reply policies/releasestrat.html]: If all that's happening is a name change, like the changes from `..._create()` to `..._new()` for some functions between 1.0.2 and 1.1.0, then the disruption caused to existing software dwarfs any benefit from the renaming. We failed to provide a smooth transition to the new names, and though the old names did not fit into the new naming pattern, they did the job.  Such renaming would have to be handled with care (the above 5 year deprecation timeline, backport of new names to old releases, ...) and the pain would have to be worth the effort.  Perhaps sometimes this will be worth the trouble, but often the old names are good enough.

**Comment by vdukhovni** [comment]
> @mlindner The company I work for uses OpenSSL heavily and we're still on 1.0.2 and looking forward to moving to 1.1.1. It's going to force a lot of code rewriting but this will be cleaner code than we had anyway and cause less bugs.

Sure, API changes may be par for the course for a single consumer of the library willing to port a handful of products.  The picture is dramatically different for OS distributions with thousands of dependent packages that need to be ported.  Since we want to see reasonably rapid uptake of the latest greatest version of OpenSSL, even as we evolve the API, we need to provide much better migration paths than we did from 1.0.2 to 1.1.0.  Yes, there are many improvements in 1.1.0, and yes we want users to port to the new way of doing things, but the transitions need to be much less painful, and it should be possible to "just recompile" old code and get it deployed with the new library, and **then** gradually port to the new way of writing the code.  One can test the need for such rewrites by building against a build Configured with "no-deprecated" or definiing `OPENSSL_API_COMPAT` to the latest version number.

File: policies/releasestrat.html
Comment (kaduk) [reply policies/releasestrat.html]: Functions that have always been documented as being not guaranteed to be API-stable would presumably get an exception (or, if later documented that way, after the same timeout as for removal).

Some other projects I work with have the luxury of an explicit export symbol list to be clear about what symbols are in the public API (intersected with the public headers, since in some cases internal symbols from one library are exposed to other libraries in the same implementation but not to the public).  Our export list is generated from the headers at present, so that just leaves us with the public headers alone, even if that leaves us in an awkward position regarding some APIs that really ought not to have existed.

File: policies/releasestrat.html
Comment (kaduk) [reply policies/releasestrat.html]: I agree with Viktor -- we definitely need to provide the compat APIs in the previous version to allow for a smooth transition. In particular, I think the following text about adding opaquification accessors should also allow adding compat APIs.
(It doesn't sound like we need to debate the actual timeout right now, but I would personally go for more like 2-3 years than 5.)

File: policies/releasestrat.html
Comment (kaduk) [reply policies/releasestrat.html]: In light of the 5-year timeline, this would not as much be "replacing" as "adding new APIs that match the naming scheme, but providing a shim layer so the old APIs still work".

File: policies/releasestrat.html
Comment (levitte) [reply policies/releasestrat.html]: Side question: how do you plan to have macros and static inline functions work against an opaque structure?

Comment (vdukhovni) [reply policies/releasestrat.html]: The text is about making the new accessors available in the **previous** release, where the structures are not yet opaque, and had no accessors.  That is users who're migrating to the new accessors should also find find them in the extant release, rather than everyone having to roll their own.

File: policies/releasestrat.html
Comment (mspncp) [inline policies/releasestrat.html]: typo: targeted

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: I put some text in to define public interface as "any function or macro declared in a public header file"

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: I clarified "modified" to say "or modified (in an API incompatible manner)"

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: I removed "security fix only mode" from that text.

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: Fixed

**Comment by mattcaswell** [comment]
New updates pushed, addressing comments so far

File: policies/releasestrat.html
Comment (tmshort) [reply policies/releasestrat.html]: Internally, we've run into instances where we've back-ported accessors and it caused problems with software packages that implemented their own accessors.

Comment (kaduk) [reply policies/releasestrat.html]: This would presumably be less of a problem if the upstream accessors were introduced in all our branches simultaneously, saving the other software from needing to ever implement their own.

File: policies/releasestrat.html
Comment (levitte) [inline policies/releasestrat.html]: The language here is confusing.

If a function is modified, it means it's name remains the same but it's signature changes. 
If it's replaced, it means a new symbol is added, and the old symbol is deprecated and removed later.

Those are two different ways of doing things, and this text seems to mix them up.

File: policies/releasestrat.html
Comment (t8m) [reply policies/releasestrat.html]: I think if this is clearly stated and followed as policy that accessors are always backported to all live branches, the 3rd party software will handle it fine. The worst thing is to keep the policy undefined - i.e. sometimes backport it and sometimes not or after a long delay, then the 3rd party implementation might really clash.

Comment (mspncp) [reply policies/releasestrat.html]: The refusal to provide an officially supported and maintained compatibility header (e.g. `<openssl/compat.h>`) to ease the transition from direct member access to accessor functions was the most significant obstacle for upgrading from 1.0.2 to 1.1.0 IMHO. It is a pity that every downstream project had to reinvent the wheel. (I say this as somebody who is currently porting our company's code from 1.0.2 to 1.1.0...)

File: policies/releasestrat.html
Comment (sam-github) [reply policies/releasestrat.html]: Renaming functions for consistency hurts all existing users of the existing name, but it helps all the future users coming to OpenSSL, and also helps the OpenSSL maintainers if it can cut down the number of questions and bug reports by confused users in the long run. I think modernizing the API is well worth doing, as long as the transition is smooth, slow and ordered. The proposed 5 year deprecation cycle seems reasonable to me.

Comment (kaduk) [reply policies/releasestrat.html]: I think we seem to all be saying that a blanket prohibition on cosmetic renaming is overkill, and that it can be safe to do with sufficient care?

File: policies/releasestrat.html
Comment (kaduk) [inline policies/releasestrat.html]: I guess there's not enough here to make it a clear win to keep the alpha/beta definitions in place until the next major release.

File: policies/releasestrat.html
Comment (tmshort) [reply policies/releasestrat.html]: Changing types to functions, specifically `int` to `ssize_t` or `size_t` is changing an interface, but is usually a source-compatible. Or even the length of the options field/API from `long` to `uin64_t`. These types of changes ("size_t'ifying") should be allowed at ABI-breaking boundaries without these restrictions.

**Comment by mspncp** [comment]
ping?

**Comment by mattcaswell** [comment]
I updated this to take account of the latest feedback. Please take another look.

File: policies/releasestrat.html
Comment (levitte) [reply policies/releasestrat.html]: I actually disagree with the removal, and rather think it should be present, but more generalised.  See #104

**Comment by mattcaswell** [comment]
I merged the updates proposed in #104 into this PR so we can see the full picture.

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: > I actually disagree with the removal, and rather think it should be present, but more generalised. See #104 

I've added the commits from #104 here and kept the more general version of this text.

File: policies/releasestrat.html
Comment (levitte) [inline policies/releasestrat.html]: The mixture of "public interface" and "original function" disturbs me, and "original function" feels linguistically clumsy.  May I suggest the following reprase?

``` html
	    <li>No existing public interface can be removed until it has been
            documented as deprecated for at least 5 years and its replacement
            has been in place in an LTS stable release. A public interface
```

**Review by levitte — CHANGES_REQUESTED** [review]

File: policies/releasestrat.html
Comment (levitte) [inline policies/releasestrat.html:177]: I'm a little confused by this...  does that mean that  will nominate MAJOR releases as LTS and include all MINOR releases therein?

Comment (mattcaswell) [reply policies/releasestrat.html:177]: An example: let's say we have releases 3.0.x and 3.1.x currently in support and 3.0.x is nominated as LTS. During the development of 4.0.0 we make a structure opaque. This means 3.0.x will have the accessor macros/functions backported to it (because it is an LTS release). They will also be backported to 3.1.x (since it is an intermediate release).

Comment (levitte) [reply policies/releasestrat.html:177]: I guess it's "feature release of the ... feature release" that throws me off. The way I have understood things is that feature releases have a new *x* in 3.*x*.0, and talking about a feature release of something that has an unchanging *x* makes that confusing...

File: policies/releasestrat.html
Comment (sam-github) [inline policies/releasestrat.html:30]: I don't understand this. Its the "only" placement that confuses me, it seems to be misplaced. You are making things more strict, so shouldn't this be:

> With this format, API/ABI compatibility will be guaranteed for all releases with the same MAJOR version number. Previously we guaranteed API/ABI compatibility only across the same MAJOR.MINOR combination.

In other words, it used to be allowed to break API from 4.5.x to 4.6.x, but now, API can be broken only from 4.x.y to 5.x.y?

File: policies/releasestrat.html
Comment (sam-github) [inline policies/releasestrat.html:44]: Perhaps this isn't the right place, but what about semver is a problem for your current LTS policies and practices? Could you point me to another conversation?

File: policies/releasestrat.html
Comment (sam-github) [reply policies/releasestrat.html:177]: At the beginning of  the document, you define MAJOR (no changes within a MAJOR can break API/ABI). But you never defined MINOR.PATCH, though you do say you aren't adopting all of semver, but that is only a statement on what they are not, not what they are. I think they should be defined.

> 3.0.x will have the accessor macros/functions backported to it

This would of course be illegal under semver. Introduction of new features requires a MINOR bump, so only changes that introduce no new features (essentially, bug fixes) can be backported to a MAJOR.MINOR.

What I _think_ you are intending after reading this paragraph and the one on top about MAJOR is this:

MAJOR.STABLE.SERIAL 

- MAJOR: API/ABI incompatible changes bump this
- STABLE: identifies a release line that shall only receive "stable" changes. occaisonally (when?) stable lines will be designated LTS. Both new features and bug fixes are allowed in a stable line (if they are considered stable)
- SERIAL: consecutive releases in a stable line get a new SERIAL number

Is something like this the intention?

You might want to avoid MINOR.PATCH, maybe its my perspective from node/npm which adheres very strongly to semver, but even using those words implies semver to me. But maybe that's not true in other software communities.

Comment (t8m) [reply policies/releasestrat.html:177]: > An example: let's say we have releases 3.0.x and 3.1.x currently in support and 3.0.x is nominated as LTS. During the development of 3.2.0 we make a structure opaque. This means 3.0.x will have the accessor macros/functions backported to it (because it is an LTS release). They will also be backported to 3.1.x (since it is an intermediate release).

I am confused now. How could you make something opaque in minor release 3.2.0 when this would break API/ABI from 3.0.0?

Comment (mattcaswell) [reply policies/releasestrat.html:177]: > I am confused now. How could you make something opaque in minor release 3.2.0 when this would break API/ABI from 3.0.0?

Oh. Sorry for the confusion. That's because my example is bad. Let's make the example this instead (I also updated the text above accordingly to save confusing anyone else):

An example: let's say we have releases 3.0.x and 3.1.x currently in support and 3.0.x is nominated as LTS. During the development of 4.0.0 we make a structure opaque. This means 3.0.x will have the accessor macros/functions backported to it (because it is an LTS release). They will also be backported to 3.1.x (since it is an intermediate release).

Comment (mattcaswell) [reply policies/releasestrat.html:177]: > MAJOR: API/ABI incompatible changes bump this
> STABLE: identifies a release line that shall only receive "stable" changes. occaisonally (when?) stable lines will be designated LTS. Both new features and bug fixes are allowed in a stable line (if they are considered stable)
> SERIAL: consecutive releases in a stable line get a new SERIAL number

Essentially, yes, although I think we will be sticking with MAJOR.MINOR.PATCH since this is already baked into the code. It could of course be changed but it's a more significant change. I think "minor" and "patch" have a meaning outside of semver.

Mostly we only allow bug fixes in a minor release - but there are exceptions. Notably for backport of accessors, and possibly "by exception" for other things.

File: policies/releasestrat.html
Comment (mspncp) [reply policies/releasestrat.html:30]: @sam-github I think you misinterpreted "only" in the sense of comparing the curren to the previous scheme (i.e, comparing the first sentence with the second). But the "only" in the first sentence referred  to the "same MAJOR version" of the same sentence.

My suggestion: either omit "only" from the first sentence (or add "only" to the second, as well).

File: policies/releasestrat.html
Comment (mspncp) [reply policies/releasestrat.html:44]: See for example https://github.com/openssl/openssl/issues/5042#issuecomment-444204003 and ff.

File: policies/releasestrat.html
Comment (mspncp) [inline policies/releasestrat.html:118]: Just a suggestion (add a "need" and get rid of the double "to be")
```
All open github issues/PRs older than 2 weeks at the time of
release need to be assessed for relevance to the version being released. 
```

File: policies/releasestrat.html
Comment (mspncp) [inline policies/releasestrat.html]: Add trailing period.

File: policies/releasestrat.html
Comment (mspncp) [inline policies/releasestrat.html]: run-checker.sh succeeds on 2 consecutive days before release

_(edited: succeds -> succeeds)_

File: policies/releasestrat.html
Comment (mspncp) [inline policies/releasestrat.html]: Side Note: then let's hope that Coverity will not be down again ;-) It is back online now, but still in readonly mode:

> Coverity Scan is back up and running in read-only mode. This means you can access your previous scan projects. We have re-built Scan using a cloud platform that will reduce dependencies and hopefully improve overall performance. We expect to be ready for new projects within the next few days once the final migration work is complete. We thank you for your patience as we work to make the platform better. For any questions in the mean time, please email sig-scan-admin@synopsys.com or visit community.synopsys.com.

File: policies/releasestrat.html
Comment (mspncp) [inline policies/releasestrat.html]: ...not to fix _the issue_ until...

File: policies/releasestrat.html
Comment (sam-github) [reply policies/releasestrat.html:30]: I think omitting only from the first sentence, or my suggested text, both work. "Only" everywhere is still open for misinterpretation.

File: policies/releasestrat.html
Comment (kaduk) [reply policies/releasestrat.html]: I'm not super-picky about which wording, but I wonder if we could get away with 3 or even 2 years.

File: policies/releasestrat.html
Comment (sam-github) [reply policies/releasestrat.html:177]: Keeping minor/patch is reasonable, and the overall scheme is, too. I do think text somewhat like the above clearly stating what minor.patch means is missing from the document.

File: policies/releasestrat.html
Comment (sam-github) [reply policies/releasestrat.html:44]: I saw a statement of "we aren't going to do it the semver way", but no rationale for why. Saying because its current policy is a head-scratcher, because the policy is changing: API breaks are no longer allowed in MINOR. Not allowing API additions in PATCH could occur at the same time, and that would be all the way to semver. LTS policies are not covered by semver, of course, those are project specific.

Node has LTS releases and uses semver. Within a LTS-designated major, if we decide to add accessors (or something of the like), we bump the minor. No big deal. Not all new features go back into LTS. They need to be considered stable, and exist in a current release for some time (for bug shakedown). Eventually LTS goes into "maintenance mode", where we don't do anything except critical bug fixes. I don't think node's stability requirements are fundamentally that different from OpenSSL, other than that your LTS are much longer (a good thing, thank you). What we do is described here, in case its of interest: https://github.com/nodejs/Release#release-plan

Anyhow, I'm done kibitzing. The policy described here will work for us as OpenSSL consumers.

File: policies/releasestrat.html
Comment (levitte) [reply policies/releasestrat.html:177]: I see no reason to go away from MAJOR.MINOR.PATCH.  Those are words for the three positions of a version number that have existed in that form for a looooong time (my memory on this takes me back to the 80's, although as MAJOR.MINOR-PATCH).  Semver is a much later effort to formalize and streamline this form of version number, but that doesn't mean that semver "owns" this version number form or the words used to describe it.

Comment (mattcaswell) [reply policies/releasestrat.html:177]: I added a description of each of MAJOR, MINOR and PATCH.

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html:30]: I dropped the word "only" here.

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html:118]: Done

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: Done

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: Done

**Comment by mattcaswell** [comment]
Fixup commit pushed addressing all comments received so far with the exception of @kaduk's suggestion of a shorter time period (2 or 3 years) for the deprecation timeout than the 5 years that this text currently says. I've not seen any discussion on that point.

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: 2 or 3 years sounds quite short to me.

Comment (levitte) [reply policies/releasestrat.html]: Regardless of length of time (I'm fine with 5 years), I'll still pick at the phrasing :wink:

File: policies/releasestrat.html
Comment (levitte) [inline policies/releasestrat.html]: This looks like a vi artifact :wink:

**Review by levitte — CHANGES_REQUESTED** [review]

File: policies/releasestrat.html
Comment (levitte) [reply policies/releasestrat.html]: I don't mind shortening the time...  but oh, this comes back to phrasing again!  I assumed that the 5 years only referred to deprecation, is that correct?  With the current phrasing, "at least 5 years" could be understood to refer both to the deprecation and the presence in an LTS stable release.

Regardless of phrasing, I'd apply much more importance in the replacement's presence in a LTS release than the length of deprecation.

Comment (mattcaswell) [reply policies/releasestrat.html]: I updated the phrasing to try and make this clearer.

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: Fixed.

File: policies/releasestrat.html
Comment (levitte) [inline policies/releasestrat.html]: I think we need to say "structure" as well.

**Review by levitte — CHANGES_REQUESTED** [review]

File: policies/releasestrat.html
Comment (levitte) [reply policies/releasestrat.html]: Oh, and just wanted to say that I like the rephrase you ended up doing.  Thank you.

File: policies/releasestrat.html
Comment (t-j-h) [inline policies/releasestrat.html]: This still remains with the problematic wording - where we labelled major.minor.fix.patch as the actual scheme and talked about "major" and "minor" releases in this separate context. That different usage should be removed in this description so we removed the opportunity for that confusion being repeated.

The simple fix is to change "Minor releases" to "Releases" and the problem goes away in this wording.

Comment (mattcaswell) [reply policies/releasestrat.html]: I applied the suggested fix

File: policies/releasestrat.html
Comment (mattcaswell) [reply policies/releasestrat.html]: I added "structure" as well. I also changed "The original function" to say "The original interface".

**Comment by mattcaswell** [comment]
Updates pushed addressing latest feedback. BTW the latest changes imply that we cannot make any more structures opaque until we have documented them as deprecated for 5 years. Do we want to say that?

**Comment by t-j-h** [comment]
No. We don't want to be boxed in to leaving a pile of things we planned to sort out and didn't get get to for a much longer time period.

**Comment by mattcaswell** [comment]
I added an exception for making structures opaque - so the first point now says:

"No existing public interface can be modified except where changes are unlikely to break source compatibility or where structures are made opaque."

**Review by levitte — APPROVED** [review]
I'm happy with this

File: policies/releasestrat.html
Comment (mlindner) [inline policies/releasestrat.html:74]: Because it's an LTS release, I would have thought 1.1.1 would be supported until the end of the terminating year (end of 2023), is that not the case?

Comment (mattcaswell) [reply policies/releasestrat.html:74]: No. 1.1.1 is supported for 5 years from its release date.

File: policies/releasestrat.html
Comment (kaduk) [inline policies/releasestrat.html]: indentation nit?

Comment (mattcaswell) [reply policies/releasestrat.html]: Fixed.

**Review by levitte — APPROVED** [review]
Still happy

**Comment by levitte** [comment]
BTW, shouldn't this be out of WIP by now?

**Comment by mattcaswell** [comment]
> BTW, shouldn't this be out of WIP by now?

Well I was leaving it in WIP until the OMC vote had concluded.

**Comment by levitte** [comment]
Ah ok

**Comment by mattcaswell** [comment]
Taken out of WIP

**Comment by mattcaswell** [comment]
This has now been approved by the OMC. I added an additional commit to update the release date. Please can I get a reconfirmation on this?

**Review by levitte — APPROVED** [review]
Yup

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**
